### ISSUE TITLE
gam print printers ssl error

### ISSUE THREAD

**Issue by dlehman83** [jay0lee/GAM#1000]
Hello, I am trying to update cloud printer shares.  Running gam print printers returns errors.  
Also running gam info printer ID returns similar results.  The ID was obtained from the cloud printer detail page in the web interface.  
I updated to the latest 4.93 as of this writing and searched the issues.  
I found an old issues stating to create a noverifyssl.txt file.  I've done this.  
Both before and after the update and text file I can still rum gam info domain, gam info user, gam print cros etc.  So I think the ssl is not the true issue.  

expected result:
A list of printers, or info about a specific printer

Actual results:
gam print printers
Traceback (most recent call last):
  File "gam.py", line 1043, in callGAPI
  File "site-packages\googleapiclient\_helpers.py", line 130, in positional_wrapper
  File "site-packages\googleapiclient\http.py", line 851, in execute
  File "site-packages\googleapiclient\http.py", line 184, in _retry_request
  File "site-packages\googleapiclient\http.py", line 165, in _retry_request
  File "gam.py", line 100, in wrapped_request_method
  File "site-packages\google_auth_httplib2.py", line 198, in request
  File "site-packages\httplib2\__init__.py", line 1953, in request
  File "site-packages\httplib2\__init__.py", line 1618, in _request
  File "site-packages\httplib2\__init__.py", line 1524, in _conn_request
  File "site-packages\httplib2\__init__.py", line 1307, in connect
  File "ssl.py", line 423, in wrap_socket
  File "ssl.py", line 870, in _create
  File "ssl.py", line 1139, in do_handshake
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:1076)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "gam.py", line 14809, in <module>
  File "gam.py", line 14329, in ProcessGAMCommand
  File "gam.py", line 3131, in doPrintPrinters
  File "gam.py", line 1068, in callGAPI
AttributeError: 'AuthorizedHttp' object has no attribute 'cache'
[1164] Failed to execute script gam

gam info printer <actual ID from web interface>
Traceback (most recent call last):
  File "gam.py", line 1043, in callGAPI
  File "site-packages\googleapiclient\_helpers.py", line 130, in positional_wrapper
  File "site-packages\googleapiclient\http.py", line 851, in execute
  File "site-packages\googleapiclient\http.py", line 184, in _retry_request
  File "site-packages\googleapiclient\http.py", line 165, in _retry_request
  File "gam.py", line 100, in wrapped_request_method
  File "site-packages\google_auth_httplib2.py", line 198, in request
  File "site-packages\httplib2\__init__.py", line 1953, in request
  File "site-packages\httplib2\__init__.py", line 1618, in _request
  File "site-packages\httplib2\__init__.py", line 1524, in _conn_request
  File "site-packages\httplib2\__init__.py", line 1307, in connect
  File "ssl.py", line 423, in wrap_socket
  File "ssl.py", line 870, in _create
  File "ssl.py", line 1139, in do_handshake
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:1076)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "gam.py", line 14809, in <module>
  File "gam.py", line 14201, in ProcessGAMCommand
  File "gam.py", line 3559, in doGetPrinterInfo
  File "gam.py", line 1068, in callGAPI
AttributeError: 'AuthorizedHttp' object has no attribute 'cache'
[12596] Failed to execute script gam

**Comment by jay0lee** [comment]
Something on your computer or network is doing SSL inspection on
www.google.com.

Jay

**Comment by dlehman83** [comment]
Yes, we just switched web filters a month ago.  I thought the text file would do the job, or that it would either work or not.  In this case some commands work and others don't.  
I bypassed the new filter and everything works as expected.  

I will reach out to their support, but is there anything I can do to make gam trust their cert?
I read in the 4.90 release notes about a CA file, but I'm not sure how to tell gam to use it.  

Thank you

**Comment by jay0lee** [comment]
noverifyssl.txt is no longer used by GAM because it's highly insecure. It
basically just told GAM to talk to anybody and screw all the protections of
SSL.

What you may be able to do is get the certificate authority file from your
proxy and point the GAM_CA_FILE environment variable at it.

Also, out of curiosity, why are you still using Cloud Print? The
recommendation these days is to move to native Chrome OS printing, it's
much more reliable.

https://support.google.com/chrome/a/answer/7276100?hl=en

Jay

**Comment by dlehman83** [comment]
I'll have to read up on that printer method, we have papercut and it links up with gcp nicely.  
What format does gam want?
This is with crt file
ERROR: [X509] no certificate or crl found (_ssl.c:4062)
This is with pem file
ERROR: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1076)

Thanks

**Comment by dlehman83** [comment]
The filter company has provided a work around.  
Thank you

**Comment by bigben386** [comment]
I recently switched to a new computer and have not been able to get the latest GAM to work. We use zscaler and do not wish to exclude google.com from ssl inspection. I have pointed the env variable to a pem file containing the zscaler root ca cert but still I get ERROR: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1076) or similar when running any gam command. Is there some special format required for the ca cert or some other way to disable ssl validation with the newer version?
